Rover/sailboat.cpp

```diff
@@ -260,6 +264,57 @@ void Sailboat::get_throttle_and_mainsail_out(float desired_speed, float &throttl
         wingsail_out *= -1.0f;
     }
 
+    //
+    // direct mast rotation control
+    // configure SERVOx_MAX/MIN such that the mast rotates +-90 degrees
+    //
+
+    if (!is_positive(desired_speed)) {
+        // rotating sails can be used to reverse, but not in this version
+        mast_rotation_out = 0.0f;
+    } else {
+        // + is wind over starboard side, - is wind over port side
+        float wind_dir_apparent_signed = degrees(rover.g2.windvane.get_apparent_wind_direction_rad());
+        float wind_dir_apparent_abs = fabsf(wind_dir_apparent_signed);
+        float wind_dir_apparent_sign = wind_dir_apparent_signed / wind_dir_apparent_abs;
+
+        // compute absolute sail rotation
+        float mast_rotation_angle = 0.0f;
+        bool mast_rotation_necessary = true;
+        if (wind_dir_apparent_abs < sail_angle_ideal) {
+            // in irons, center the sail.
+            mast_rotation_angle = 0.0f;
+        } else if (wind_dir_apparent_abs < (90.0f + sail_angle_ideal)) {
+            // use sail as a lift device, at ideal angle of attack
+            mast_rotation_angle = (wind_dir_apparent_abs - sail_angle_ideal);
```

Comment:
```suggestion
            mast_rotation_angle = wind_dir_apparent_abs - sail_angle_ideal + pid_offset;
```
Need to add back in the ability to sheet out if were tipping over.

Reply:
Yes. I need to account for `pid_offset`.

`mast_rotation_angle` is in degrees though, and as I recall, `pid_offset` is in percent, so I don't think you can add them here. (https://discord.com/channels/674039678562861068/797946854792757259/829132730327105576)

From reading `AR_AttitudeControl::get_sail_out_from_heel()` I _think_ the result is always zero or negative. If that is the case, you'd want to _subtract_ it from mast rotation. You want to _decrease_ the AoA, i.e. bring the sail closer to the wind to depower it. Let me know if I have this backwards.

Reply:
I think you're right to use `pid_offset` here, only for upwind points of sail. What is this value a percent of? If it a fraction of available power that it wants me to use, like "we're heeling too much, reduce power by 30%?" If this is how it works, it may be easier to _multiply_ `sail_angle_ideal` by this fraction instead of _adding_ this value in the end. This assumes linearity between angle of attack and lift, which is shaky... or just makes this shaky assumption explicit.

Reply:
`pid_offset` is a none-dimensional output of the PID controller, you can use it however you like. You could add it to the scaled value at the end to be more consistent with the mainsail.  We add it to mainsail_base, so I think adding it here is correct too. The assumption here is of course that you only have one type of sail fitted at a time, the PIDs won't sheet out each type a equal amount.

Reply:
On a wider note, I keep tripping over the idea that AP sets sails in response to wind changes instead of adjusting the rudder. On a real boat, even a small one, sheeting is expensive in terms of both work and time, especially when there are multiple sheets. If you're about to capsize, you'd steer to fall off the wind ("tiller towards trouble"), not try to sheet out. I have to remind myself that Rover's pedigree is in powered vehicles that simply push through disturbances, but it messes me up every time I try to understand the wider logic of the code. The same happens to me when dealing with Plane. By default, Plane wants to use throttle to control airspeed, and then respond to changes in altitude with elevator, which is opposite from how you fly a full-size plane.

Reply:
Yeah, you would, that is very hard to program. In reality you would almost always do a combination of both and sheet ease and a course change. Depending on the situation and wind speed and point of sail you might do more of one or the other. Plane has the TECS system, and all its parameters to decide that, we need something similar for this. 

Its only sail at the moment because its much easier to change the sail output than to break into the steering output and clutter the code from normal rover users. We also have to stay in our box, if the user has set a heading we cant just sail at a different heading and not sheet out. Really we could only do it properly in modes where we have a target location and are free to pick the best heading to get us there.

Reply:
It's fine as an approximation, and on little boats with full-time winch servos and relatively large batteries it works fine. (May not work as well in terms of battery capacity on Microtransat which was my original motivation for this boat). What you refer to as "a combination of both" is feed forward. If you want your airplane to climb, you increase demanded throttle directly, and use FF to tell the system it's going to need to start pulling up to maintain airspeed.

> ...break into the steering output and clutter the code from normal rover users...
> If the user has set a heading we cant just sail at a different heading...

That's the power vehicle pedigree, and I respect that. But we're not talking about choosing best course for destination and wind here. We're talking about preventing a capsize (plane equivalent is avoiding stall). In any case, this part of the conversation if outside the scope of my PR. If you ever decide to look at making rudder the primary control without messing up the rest of the codebase, I'm happy to be a sounding board.

What I need here is for you to tell me about `pid_offest`. Does it have a sign? Is it always positive? I understand that when you wrote this originally, there was a very strong assumption that a sheet works the same port and starboard, but it would be wonderful if you could document your assumptions and invariants either here or in `AR_AttitudeControl.h`. I can see that you set `p` and `i` to zero if they are positive, but not `d` or `ff`. Then you multiply the sum by -1.0f, calling it "constrain and return final output." It is probably my lack of familiarity with AP PID controllers, but by now I'm completely lost about what range of values to expect. Then in `sailboat.cpp` you multiply this value by 100 and constrain it to [0.0f..100.0f], and I have no idea what I'm looking at. Help me out?

Reply:
Yeah, you will have to swap the sign of the PID offset depending on the tack. I'm not sure off the top of my head which way round.

Reply:
@IamPete1, you're answering a different question from the one I'm asking.

Reply:
19ff950c7be91ef2c5d3bd42cbf74b0160d022ad should address this feedback.